lib/src/facts/aix/kernel_resolver.cc

```diff
@@ -0,0 +1,70 @@
+#include <internal/facts/aix/kernel_resolver.hpp>
+#include <internal/util/regex.hpp>
+#include <facter/facts/collection.hpp>
+#include <leatherman/logging/logging.hpp>
+#include <facter/execution/execution.hpp>
+#include <facter/util/file.hpp>
+#include <boost/regex.hpp>
+
+using namespace facter::util;
+using namespace facter::execution;
+using namespace std;
+
+namespace facter { namespace facts { namespace aix   {
+
+    kernel_resolver::data kernel_resolver::collect_data(collection& facts)
+    {
+        data result;
+        const auto regex = boost::regex("(\\d+-\\d+-\\d+)_SP.*Service Pack");
+        vector<string> versions;
+
+        // We reimplement part of oslevel here. We parse the rml cache
+        // and use that to determine the full AIX system version. If
+        // the rml cache doesn't exist, we just call out to oslevel to
+        // regenerate it (since we're gonna be slow in that case
+        // anyway).
+        bool success;
+        bool tried_oslevel = false;
+        do {
+            success = file::each_line("/tmp/.oslevel.datafiles/.oslevel.rml.cache", [&](string& line) {
+                string value;
+                re_search(line, regex, &value);
+                if (!value.empty()) {
+                    versions.push_back(value);
+                }
+                return true;
+            });
+
+            // If we can't open the file we assume it just doesn't
+            // exist, and that oslevel will regenerate it.
+            if (!success) {
+                if (tried_oslevel) {
+                    // If we already tried running oslevel the last
+                    // time through, we can probably assume it won't
+                    // work the second time. We just bail at this
+                    // point.
+                    LOG_WARNING("Could not parse rml cache after several tries: kernel facts are unavailable. Try running 'oslevel -s' to debug.");
+                    return result;
+                }
+                bool oslevel_success;
+                string out;
+                string err;  // unused, but needed for tuple unpacking
+                std::tie(oslevel_success, out, err) = execute("/usr/bin/oslevel", {"-s"}, 0, { execution_options::trim_output, execution_options::redirect_stderr_to_stdout, execution_options::merge_environment });
+                if (!oslevel_success) {
+                    LOG_WARNING("oslevel failed: %1%: kernel facts are unavailable", out);
+                    return result;
+                }
+                tried_oslevel = true;
+            }
+        } while (!success);
+
+        sort(versions.rbegin(), versions.rend());
```

Comment:
This code is from Whopper, and I didn't dig into it in too much depth. It's probably a pretty faithful recreation of the bash, which doesn't have such niceties. I'll massage it a bit more